modules/firehose/main.tf

```diff
@@ -141,12 +150,93 @@ resource "aws_iam_role_policy" "firehose_to_http_metric_policy" {
            "Resource": [
                "${aws_cloudwatch_log_group.firehose_loggroup.arn}"
            ]
+        },
+        {
+          "Effect": "Allow",
+          "Action": [
+              "lambda:InvokeFunction",
+              "lambda:GetFunctionConfiguration"
+          ],
+          "Resource": "${aws_lambda_function.lambda_processor.arn}:*"
         }
     ]
 }
 EOF
 }
 
+data "aws_iam_policy_document" "lambda_assume_role" {
+  statement {
+    effect = "Allow"
+
+    principals {
+      type        = "Service"
+      identifiers = ["lambda.amazonaws.com"]
+    }
+
+    actions = ["sts:AssumeRole"]
+  }
+}
+
+resource "aws_iam_role" "lambda_iam" {
+  name               = "${var.firehose_stream}-transform-lambda-iam"
+  tags               = local.tags
+  assume_role_policy = data.aws_iam_policy_document.lambda_assume_role.json
+}
+
+resource "aws_iam_role_policy" "lambda_iam_policy" {
+  name   = "${var.firehose_stream}-transform-lambda-iam"
+  role   = aws_iam_role.lambda_iam.id
+  policy = <<EOF
+{
+  "Version": "2012-10-17",
+  "Statement": [
+      {
+          "Action": [
+              "tag:GetResources",
+              "ec2:DescribeTransitGateway*",
+              "ec2:DescribeTags",
+              "ec2:DescribeRegions",
+              "ec2:DescribeInstances",
+              "dms:DescribeReplicationTasks",
+              "dms:DescribeReplicationInstances",
+              "apigateway:GET"
+          ],
+          "Effect": "Allow",
+          "Resource": "*",
+          "Sid": ""
+      },
+      {
+          "Action": [
+              "logs:PutLogEvents",
+              "logs:CreateLogStream",
+              "logs:CreateLogGroup"
+          ],
+          "Effect": "Allow",
+          "Resource": "arn:aws:logs:*:*:*",
+          "Sid": ""
+      }
+  ]
+}
+EOF
+}
+
+resource "aws_cloudwatch_log_group" "loggroup" {
+  name              = "/aws/lambda/${aws_lambda_function.lambda_processor.function_name}"
+  retention_in_days = var.cloudwatch_retention_days
+  tags              = local.tags
+}
+
+resource "aws_lambda_function" "lambda_processor" {
+  s3_bucket = "cx-cw-metrics-tags-lambda-processor-dev-us-east-2" # TODO: Support all regions
```

Comment:
Since we have plans to support more than one region, my advice is to let this code be prepared.
We can have a variable with the region name and concatenate it with the bucket name, or if the bucket could be customised we should also provide a variable for that.